### DISCUSSION TITLE
IP Adapters: Negative image

### DISCUSSION THREAD

**Discussion by asomoza** [huggingface/diffusers#7167] [Ideas]
I'm starting this discussion to document and share some examples of this technique with IP Adapters.

First of all, this wasn't my initial idea, so thanks to @cubiq and his repository https://github.com/cubiq/ComfyUI_IPAdapter_plus. It is from here and his comments that I got started playing with this idea. AFAIK there's no mention about this in the official repository or the paper.

For this discussion I'm only applying noise and images that can be generated automatically, there's a lot more that can be done with manual intervention but those would be better with an UI.

Lets start with an initial image without prompts, I'm using the IP Adapter PLUS with a 1.0 scale, all settings are the same with a fixed seed, so this is the base where I start:

|source|result|
|---|---|
|![20240229150812](https://github.com/huggingface/diffusers/assets/5442875/901d8bd8-7a59-4fe7-bda1-a0e0d6c7dffd)|![20240229163923_normal](https://github.com/huggingface/diffusers/assets/5442875/3432e25a-ece6-45f4-a3f4-fca354f40b5b)|

The quality isn't that great right now. I'm using the latest juggernaut model and what I'm trying to achieve is a more cinematic result and to change the initial setting of the image, so I'll add this prompts:

```py
prompt = "cinematic photo of a cyborg in the city, 4k, high quality, intricate, highly detailed"
negative_prompt = "blurry, smooth, plastic"
```

The result with this prompt is like this:

|without IP Adapter| With IP Adapter|
|---|---|
|![20240301051842](https://github.com/huggingface/diffusers/assets/5442875/bc1f24ea-b890-4db5-8155-47fc7160c10b)|![20240229180838_prompt](https://github.com/huggingface/diffusers/assets/5442875/1f60e2cb-c114-44a8-9189-9551465b3755)|

The quality still isn't that great, also even with the prompt there's almost no city in the image. So now I'll start with the negative image of IP Adapters to try to make it better, first I'll start with mandelbrot noise, one normal and one inverted.

|mandelbrot|mandelbrot inverted|
|---|---|
|![mandelbrot](https://github.com/huggingface/diffusers/assets/5442875/e65a35c2-2469-45cd-bdc5-fb884b27cb16)|![mandelbrot_inverted](https://github.com/huggingface/diffusers/assets/5442875/78923234-f3fc-4a1a-bb7a-333c634c1f72)|

|normal result|inverted result|
|---|---|
|![20240229181056_mandelbrot](https://github.com/huggingface/diffusers/assets/5442875/e828cab7-fd44-4688-a736-81077415e8f4)|![20240229181135_mandelbrot_inverted](https://github.com/huggingface/diffusers/assets/5442875/59c42de2-1283-475f-9569-87162ff86318)|

I like it better the result with the inverted mandelbrot, but still it doesn't have that much of a city so I had to lower the scale of the IP Adapter to 0.5, but with that and without controlnet I lose the composition position and pose of the cyborg.

Even without that I still think it looks good so here's the result:

<p align="center">
<img src="https://github.com/huggingface/diffusers/assets/5442875/30a33700-a343-4eed-a020-eedf1c72cd79" alt="IP Adapter with prompt" width="500" height="500">
</p>

To compare the results with another type of noise, I did gaussian:
||||
|---|---|---|
|![gaussian_5](https://github.com/huggingface/diffusers/assets/5442875/c85e2827-e991-4ed0-9ae6-7414662b9352)|![gaussian_15](https://github.com/huggingface/diffusers/assets/5442875/a87b1b02-7296-4d36-a121-587fbd080e20)|![gaussian_30](https://github.com/huggingface/diffusers/assets/5442875/530c142c-841d-4e6e-94f7-a7bc7208c05e)|

Results:
||||
|---|---|---|
|![20240229182843_gaussian_5_str05](https://github.com/huggingface/diffusers/assets/5442875/ebd79452-c86c-4c87-a4c7-400d59e0b7a5)|![20240229182911_gaussian_15_str05](https://github.com/huggingface/diffusers/assets/5442875/efbde387-b31d-49ec-a5e1-677650b7c2c6)|![20240229182911_gaussian_30_str05](https://github.com/huggingface/diffusers/assets/5442875/a5114f9e-3a7d-4cef-b375-9340fae21652)|

So all this got me thinking, how about if I start feeding it other types of images as negatives, for example a blurred image:

|blurred image|result scale 0.5|result scale 1.0|
|---|---|---|
|![gaussian_blur_30](https://github.com/huggingface/diffusers/assets/5442875/71b0ab11-d4b3-40ac-ba80-5ef77486eac6)|![20240301014729](https://github.com/huggingface/diffusers/assets/5442875/84a2c0b1-cc66-43f4-9567-dd7cacb818c3)|![20240301014814](https://github.com/huggingface/diffusers/assets/5442875/dd45ee2c-662c-4ac1-bc2e-1428cb2e0144)|

So I got a really sharp image of the cyborg, with a scale of 0.5 it didn't have any of the original image left so I used a scale of 1.0.

This again got me thinking, if a blurred image makes it sharper, what about colors, so I tested it with passing the isolated color channels of the image as negatives:

|red|green|blue|
|---|---|---|
|![red](https://github.com/huggingface/diffusers/assets/5442875/9d6ecfb4-d706-41de-be3e-2c610b9ae539)|![green](https://github.com/huggingface/diffusers/assets/5442875/bc8d9423-f226-4c3b-a920-e8b59cdf6a41)|![blue](https://github.com/huggingface/diffusers/assets/5442875/98f4444b-423a-4a47-a53b-96b8f817a62e)|
|![20240301050326](https://github.com/huggingface/diffusers/assets/5442875/43df5885-f5b4-46b6-8676-00f8a64f7177)|![20240301050418](https://github.com/huggingface/diffusers/assets/5442875/9a03e2f2-cd18-4380-9401-34f2d83dccad)|![20240301050457](https://github.com/huggingface/diffusers/assets/5442875/07117ea4-73fb-4198-8bf1-ac2bb7399186)|

That wasn't the result I was expecting but I like the blue and green ones, so as a final test, since I liked those two and I wanted a sharper image, I did a mix of those three:

|blue + green|low gaussian blur|final negative image|
|---|---|---|
|![blue_green](https://github.com/huggingface/diffusers/assets/5442875/40d8a2d1-bf06-43b8-86a6-d56c4b10b5ae)|![gaussian_blur](https://github.com/huggingface/diffusers/assets/5442875/b501cc96-d5dc-455b-b268-ffea49c7f0d7)|![blue_green_gaussian](https://github.com/huggingface/diffusers/assets/5442875/028d1545-f884-4b2b-a7bb-5c55edf92b17)|

With this I got the image I was looking for, still need some inpainting to fix details, but IMO it looks really good to be generated with just a single IP Adapter Image:

![20240301020459](https://github.com/huggingface/diffusers/assets/5442875/736fd15a-36ba-40c0-a7d8-6ec1ac26f788)

Without masks and controlnet, the use of the IP Adapter here is like a creative start (instead of using prompts you use an image to feed the model with your initial idea). If you want to do the same but preserving more of the initial image you'll need to use them.

**Comment by asomoza** [comment]
For reference, this is the code I used:

```python
import torch

from diffusers import AutoencoderKL, DPMSolverMultistepScheduler, StableDiffusionXLPipeline
from diffusers.models import ImageProjection
from diffusers.utils import load_image


def encode_image(
    image_encoder,
    feature_extractor,
    image,
    device,
    num_images_per_prompt,
    output_hidden_states=None,
    negative_image=None,
):
    dtype = next(image_encoder.parameters()).dtype

    if not isinstance(image, torch.Tensor):
        image = feature_extractor(image, return_tensors="pt").pixel_values

    image = image.to(device=device, dtype=dtype)
    if output_hidden_states:
        image_enc_hidden_states = image_encoder(image, output_hidden_states=True).hidden_states[-2]
        image_enc_hidden_states = image_enc_hidden_states.repeat_interleave(num_images_per_prompt, dim=0)

        if negative_image is None:
            uncond_image_enc_hidden_states = image_encoder(
                torch.zeros_like(image), output_hidden_states=True
            ).hidden_states[-2]
        else:
            if not isinstance(negative_image, torch.Tensor):
                negative_image = feature_extractor(negative_image, return_tensors="pt").pixel_values
            negative_image = negative_image.to(device=device, dtype=dtype)
            uncond_image_enc_hidden_states = image_encoder(negative_image, output_hidden_states=True).hidden_states[-2]

        uncond_image_enc_hidden_states = uncond_image_enc_hidden_states.repeat_interleave(num_images_per_prompt, dim=0)
        return image_enc_hidden_states, uncond_image_enc_hidden_states
    else:
        image_embeds = image_encoder(image).image_embeds
        image_embeds = image_embeds.repeat_interleave(num_images_per_prompt, dim=0)
        uncond_image_embeds = torch.zeros_like(image_embeds)

        return image_embeds, uncond_image_embeds


@torch.no_grad()
def prepare_ip_adapter_image_embeds(
    unet,
    image_encoder,
    feature_extractor,
    ip_adapter_image,
    do_classifier_free_guidance,
    device,
    num_images_per_prompt,
    ip_adapter_negative_image=None,
):
    if not isinstance(ip_adapter_image, list):
        ip_adapter_image = [ip_adapter_image]

    if len(ip_adapter_image) != len(unet.encoder_hid_proj.image_projection_layers):
        raise ValueError(
            f"`ip_adapter_image` must have same length as the number of IP Adapters. Got {len(ip_adapter_image)} images and {len(unet.encoder_hid_proj.image_projection_layers)} IP Adapters."
        )

    image_embeds = []
    for single_ip_adapter_image, image_proj_layer in zip(
        ip_adapter_image, unet.encoder_hid_proj.image_projection_layers
    ):
        output_hidden_state = not isinstance(image_proj_layer, ImageProjection)
        single_image_embeds, single_negative_image_embeds = encode_image(
            image_encoder,
            feature_extractor,
            single_ip_adapter_image,
            device,
            1,
            output_hidden_state,
            negative_image=ip_adapter_negative_image,
        )
        single_image_embeds = torch.stack([single_image_embeds] * num_images_per_prompt, dim=0)
        single_negative_image_embeds = torch.stack([single_negative_image_embeds] * num_images_per_prompt, dim=0)

        if do_classifier_free_guidance:
            single_image_embeds = torch.cat([single_negative_image_embeds, single_image_embeds])
            single_image_embeds = single_image_embeds.to(device)

        image_embeds.append(single_image_embeds)

    return image_embeds


vae = AutoencoderKL.from_pretrained(
    "madebyollin/sdxl-vae-fp16-fix",
    torch_dtype=torch.float16,
).to("cuda")

pipeline = StableDiffusionXLPipeline.from_pretrained(
    "RunDiffusion/Juggernaut-XL-v9",
    torch_dtype=torch.float16,
    vae=vae,
    variant="fp16",
).to("cuda")

pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config)
pipeline.scheduler.config.use_karras_sigmas = True

pipeline.load_ip_adapter(
    "h94/IP-Adapter",
    subfolder="sdxl_models",
    weight_name="ip-adapter-plus_sdxl_vit-h.safetensors",
    image_encoder_folder="models/image_encoder",
)
pipeline.set_ip_adapter_scale(0.7)

ip_image = load_image("source.png")
negative_ip_image = load_image("noise.png")

image_embeds = prepare_ip_adapter_image_embeds(
    unet=pipeline.unet,
    image_encoder=pipeline.image_encoder,
    feature_extractor=pipeline.feature_extractor,
    ip_adapter_image=[[ip_image]],
    do_classifier_free_guidance=True,
    device="cuda",
    num_images_per_prompt=1,
    ip_adapter_negative_image=negative_ip_image,
)


prompt = "cinematic photo of a cyborg in the city, 4k, high quality, intricate, highly detailed"
negative_prompt = "blurry, smooth, plastic"

image = pipeline(
    prompt=prompt,
    negative_prompt=negative_prompt,
    ip_adapter_image_embeds=image_embeds,
    guidance_scale=6.0,
    num_inference_steps=25,
    generator=torch.Generator(device="cpu").manual_seed(1556265306),
).images[0]

image.save("result.png")
```

**Comment by yiyixuxu** [comment]
this is awesome :)

**Comment by cubiq** [comment]
monochrome noise doesn't work very well in my tests. An interesting way of adding noise might be as follow

```
factor = 0.5 # 0.0....1.0
mask = (torch.rand_like(image) < factor).float()
noise = torch.rand_like(image)
noise = torch.zeros_like(image) * (1-mask) + noise * mask
```
you can further calibrate the effect by multiplying the noise by the factor again at the end for an even more subtle effect.
This way you have a very gradual conditioning where at 0.01 you get a very light effect and it starts to get more impactful as you increase the noise factor.

PS: it is important to remember that the noise generation needs to be seeded if you want repeatable results

**Reply by asomoza** [reply]
yeah, I forgot to add that note. I normally do this with the tensors instead of pillow images and as parameters instead of images, so for example the type of noise and strength (forgot the seed thank you) as arguments.

I was toying with the idea of a factor or maybe add opacity if they're images. I will try your example and CFG strategies that work best with this.

**Comment by yiyixuxu** [comment]
what a great post! thank you for sharing this

I'm adding the link to the initial discussion on the IP adapter negative image here https://github.com/huggingface/diffusers/issues/6318#issuecomment-1920909332

to summarize based on my understanding: the negative image allows you to generate images that have more variation from the original ip-adapter image; you can also do that by lowering the `ip_adapter_scale, but with the negative image you can have more control over the generation e.g. you can preserve more of the composition of the original image ad only lose the details you want to modify etc 

super neat technique :)

**Reply by asomoza** [reply]
yes but if the noise is over the same image and in a low quantity sometimes it improves the quality of the generation without changing the composition of the image but IMO this is relative to the person that uses it and sometimes it loses some details too.

In this case I was using it as you described, there's some other tests I need to do like lower the influence factor of the noise and use other kinds of negative images.

**Reply by sayakpaul** [reply]
Sorry if this is a dumb question. Conceptually, how does the structure of the negative noise is incorporated during the generation? Is it more like "we don't want any resemblance from the negative image in the final generated image"?

**Reply by asomoza** [reply]
this is all my speculation, but yes and no, because the image you pass as a negative also has parts of the positive image so is more like a guidance, if we transform the image prompt to a text prompt (simplified a lot) it would be like this:

positive prompt = "apple, pear, grapes"
negative prompt = "apple"

if you do that, you will still get an apple in the image most of the time but if you do this:

positive prompt = "apple, pear, grapes"
negative prompt = "red apple"

most of the time you won't get a red apple but a green apple instead.

So if we pass a negative noise image we're telling the model to do the same image but with less noise, if the negative image has a lot of noise, the result image is smoother, but when we pass an image with very little noise, the image gets more details somehow, but we can't really know what the model is interpreting with those embeddings.

Using this simple function to generate the noise:

```py
def generate_noise_image(image):
    torch.manual_seed(noise_seed)
    generated_noise = torch.randn(*image.shape, device=image.device) * noise_factor
    noise_image = image + generated_noise
    noise_image = torch.clamp(noise_image, 0, 1)

    return noise_image
```

This are the results:

|source|no noise|0.01 noise|0.1 noise|0.5 noise|1.0 noise|
|---|---|---|---|---|---|
|![20240304054417](https://github.com/huggingface/diffusers/assets/5442875/46412d6f-917f-42fb-b1d7-befb40f51d99)|![20240304070545_no_noise](https://github.com/huggingface/diffusers/assets/5442875/96671edf-8894-4f46-936b-20c3d811bfcb)|![20240304070735_0 01](https://github.com/huggingface/diffusers/assets/5442875/7e8e5b83-e1df-478a-b716-7fdad8a6e12b)|![20240304070816_0 1](https://github.com/huggingface/diffusers/assets/5442875/d070196a-f84b-4a16-8193-0b0037e7ff6f)|![20240304070843_0 5](https://github.com/huggingface/diffusers/assets/5442875/81b03b76-8ade-42b6-a884-6f5bc1a90590)|![20240304070937_0](https://github.com/huggingface/diffusers/assets/5442875/79418ca9-83b7-43a4-8afc-c9f2a1aa03d7)|

The 0.01 image has a more details and the shadows are better, so if you ask me, I like a lot more the result with 0.01 noise.

This is an animation with everything fixed except the noise factor:

![noise_demo](https://github.com/huggingface/diffusers/assets/5442875/3e65e9a1-bc36-48e3-9d9a-3e14aaa4fdba)

It's not a smooth transition, for each step of the noise, the image changes, so if you want to just improve the image without changing that much of it, you'll have to use a very low noise factor. 

I'm guessing that what the model interprets is something like "do this image but not exactly the same image" and it ends adding more details or shadows to achieve this.

I use this as an additional parameter to change the generation, the same as the people that use weird words in the negative prompt and sometimes they get good unexpected results.

**Reply by sayakpaul** [reply]
Thanks a lot for explaining!

The images in the comparative table you showed above -- are those images just right after adding the noise? How were those generated? 

In the animation -- on what aspects do the two animations differ from one another?

**Reply by asomoza** [reply]
> The images in the comparative table you showed above -- are those images just right after adding the noise? How were those generated?

They have everything fixed except the noise, I generated them using the function above and monkey patching the encode_image:

```py
def encode_image(self, image, device, num_images_per_prompt, output_hidden_states=None):
    dtype = next(self.image_encoder.parameters()).dtype
    image = self.feature_extractor(image, return_tensors="pt").pixel_values
    image = image.to(device=device, dtype=dtype)

    # generate noise image
    image_enc_hidden_states = self.image_encoder(image, output_hidden_states=True).hidden_states[-2]
    image_enc_hidden_states = image_enc_hidden_states.repeat_interleave(num_images_per_prompt, dim=0)

    noise_image = generate_noise_image(image)
    uncond_image_enc_hidden_states = self.image_encoder(noise_image, output_hidden_states=True).hidden_states[-2]

    # zero tensor
    # uncond_image_enc_hidden_states = self.image_encoder(
    #     torch.zeros_like(image), output_hidden_states=True
    # ).hidden_states[-2]
    uncond_image_enc_hidden_states = uncond_image_enc_hidden_states.repeat_interleave(num_images_per_prompt, dim=0)
    return image_enc_hidden_states, uncond_image_enc_hidden_states
```
I still want to test more variations and I'll post the whole code if someone wants to play with it.

> In the animation -- on what aspects do the two animations differ from one another?

The animation is just a side by side comparison of the `result image` (left) and the `noise image` (right) saved with `torchvision.utils.save_image` that was used to generate it. The noise factors of the animation were generated like this:

```py
noise_factors = np.linspace(0.0, 1.0, num=50)
```

**Reply by asomoza** [reply]
Just a follow up, if I use an UI with all of this plus t2i adapters, ip adapters masking and prompts, this is the result:

![20240305002208](https://github.com/huggingface/diffusers/assets/5442875/8ee207b8-5157-485a-a015-8586f158feee)

**Reply by cubiq** [reply]
if you are going to use just noise you can try to send directly noisy uncond embeds. that saves you encoding the image. in which case you can use the code I posted earlier

**Reply by asomoza** [reply]
oh yeah, thanks, I totally missed that lol. I'm going to test now with that code and playing with multiplying the factor after and maybe do a combination with gradually lowering the CFG.

**Reply by cubiq** [reply]
does diffusers have a CFG-rescale strategy? if so, go for it, otherwise you need to drastically lower the CFG

**Reply by sayakpaul** [reply]
We do:

https://github.com/huggingface/diffusers/blob/687bc2772721af584d649129f8d2a28ca56a9ad8/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L59
